rdtools/system_analysis.py

```diff
@@ -0,0 +1,587 @@
+        case: 'sensor' or 'clearsky' which filtering protocol to apply. Affects whether filtering.csi_filter()
+               is used and whether result is stored in self.sensor_filter or self.clearsky_filter)
+
+        Returns
+        -------
+        None
+        '''
+        bool_filter = True
+
+        if case == 'sensor':
+            poa = self.poa
+            cell_temp = self.cell_temperature
+        if case == 'clearsky':
+            poa = self.clearsky_poa
+            cell_temp = self.clearsky_cell_temperature
+
+        if 'normalized_filter' in self.filter_params.keys():
+            f = filtering.normalized_filter(normalized, **self.filter_params['normalized_filter'])
+            bool_filter = bool_filter & f
+        if 'poa_filter' in self.filter_params.keys():
+            if poa is None:
+                raise ValueError('poa must be available to use poa_filter')
+            f = filtering.poa_filter(poa, **self.filter_params['poa_filter'])
+            bool_filter = bool_filter & f
+        if 'tcell_filter' in self.filter_params.keys():
+            if cell_temp is None:
+                raise ValueError('Cell temperature must be available to use tcell_filter')
+            f = filtering.tcell_filter(cell_temp, **self.filter_params['tcell_filter'])
+            bool_filter = bool_filter & f
+        if 'clip_filter' in self.filter_params.keys():
+            if self.pv_power is None:
+                raise ValueError('PV power (not energy) is required for the clipping filter. Either omit the clipping filter,'
+                                 'provide PV power at instantiation, or explicitly assign system_analysis.pv_power.')
+            f = filtering.clip_filter(self.pv_power, **self.filter_params['clip_filter'])
+            bool_filter = bool_filter & f
+        if 'ad_hoc_filter' in self.filter_params.keys():
+            if self.filter_params['ad_hoc_filter'] is not None:
+                bool_filter = bool_filter & self.filter_params['ad_hoc_filter']
+        if case == 'clearsky':
+            if self.poa is None or self.clearsky_poa is None:
+                raise ValueError('Both poa and clearsky_poa must be available to do clearsky filtering with csi_filter')
+            f = filtering.csi_filter(self.poa, self.clearsky_poa, **self.filter_params['csi_filter'])
+            bool_filter = bool_filter & f
+
+        if case == 'sensor':
+            self.sensor_filter = bool_filter
+        elif case == 'clearsky':
+            self.clearsky_filter = bool_filter
+
+    def aggregate(self, normalized, insolation):
+        '''
+        Returns insolation-weighted normalized PV energy and the associated aggregated insolation
+
+        Parameters
+        ----------
+        normalized: Pandas Times Series (numeric), normalized pv energy time series
+        insolation: Pandas Time Series (numeric), Time series of insolation associated
+                    with each normalize_energy point
```

Comment:
should `normalize_energy` be `normalized`?

Reply:
yep, fixed